### PR TITLE
Fix for #348 - IGListStackedSectionController's children need to know numberOrItems before didUpdate is called

### PR DESCRIPTION
**Pull request by jeffbailey**
## Changes in this pull request
The reloadData method in IGListStackedSectionController was being called too soon.  It was moved from the constructor to the `didUpdateToObject` method.  Tests were updated to reflect the change.

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [ ] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by jessesquires** [comment]
Nice. Thanks @jeffbailey ! 💯

File: Tests/IGListStackSectionControllerTests.m [hunk 1]
```diff
@@ -85,8 +85,9 @@ - (void)test_whenInitializingStack_thatNumberOfItemsMatches {
     section4.items = 1;
 
     IGListStackedSectionController *stack = [[IGListStackedSectionController alloc] initWithSectionControllers:@[section1, section2, section3, section4]];
-
-    XCTAssertEqual([stack numberOfItems], 6);
+    [stack performBatchAnimated:false updates:^{} completion:^(BOOL finished) {
+        XCTAssertEqual([stack numberOfItems], 6);
```
Comment (rnystrom) [inline Tests/IGListStackSectionControllerTests.m]: These tests use [IGListAdapterUpdater](https://github.com/Instagram/IGListKit/blob/master/Tests/IGListStackSectionControllerTests.m#L57) which asynchronously collect updates during a runloop turn, so this assert wont execute w/out an expectation+wait setup. Here's an [example](https://github.com/Instagram/IGListKit/blob/master/Tests/IGListStackSectionControllerTests.m#L766-L776). Same goes for all the asserts in this change that got moved into the completion block.

A really easy way to check that the block executes is:

```objc
__block BOOL executed = NO;
[adapter performUpdates:^{
  executed = YES;
}];
XCTAssertTrue(executed);
```

Comment (jeffbailey) [inline Tests/IGListStackSectionControllerTests.m]: Yep, it should have dawned on me I made the changed tests asynchronous :).  The issue with these particular tests is that they don't follow the normal pattern of the other tests because they don't use the adapter.  I don't know without digging further why they don't use the adapter or if they can be changed easily.  Also, my proposed solution wouldn't work anyway since the completion block never gets called because there's no CollectionContext set on the section controller.  One easy way to use to fix the broken tests would be to use the following pattern:

```
- (void)test_whenInitializingStack_thatNumberOfItemsMatches {
    IGListTestSection *section1 = [[IGListTestSection alloc] init];
    section1.items = 2;
    IGListTestSection *section2 = [[IGListTestSection alloc] init];
    section2.items = 3;
    IGListTestSection *section3 = [[IGListTestSection alloc] init];
    section3.items = 0;
    IGListTestSection *section4 = [[IGListTestSection alloc] init];
    section4.items = 1;

    IGListStackedSectionController *stack = [[IGListStackedSectionController alloc] initWithSectionControllers:@[section1, section2, section3, section4]];
    [stack performSelectorOnMainThread:@selector(reloadData) withObject:nil waitUntilDone:TRUE];
    XCTAssertEqual([stack numberOfItems], 6);
}
```

Basically just call the private reloadData method using performSelector.  Let me know what you think the best way to proceed would be.

Comment (jeffbailey) [inline Tests/IGListStackSectionControllerTests.m]: @rnystrom - let me know what you think about just using:
`    [stack performSelectorOnMainThread:@selector(reloadData) withObject:nil waitUntilDone:TRUE];
`
to get the unit tests working.

Comment (rnystrom) [inline Tests/IGListStackSectionControllerTests.m]: @jeffbailey oh! I missed that. Good call.

Instead of doing the `performSelector` hack, how about we expose `reloadData` in the internal header and call it directly?

https://github.com/Instagram/IGListKit/blob/master/Source/Internal/IGListStackedSectionControllerInternal.h

The header is already imported to the test.

**Comment by jessesquires** [comment]
Friendly ping ✨ any updates here?

@rnystrom if this is good, we can follow up with a changelog entry separately.

**Comment by jeffbailey** [comment]
The internal header is a great option.  I'll get this done today.  Thanks @rnystrom

**Comment by rnystrom** [comment]
@jeffbailey looks great! Very last thing is a quick note in the `CHANGELOG.md` file about this fix and we're g2g. It should go under the `2.2.0` list.
